Fungible tokens tutorial (#753)
Concordium/concordium.github.io: source/mainnet/smart-contracts/tutorials/fungible-tokens/mint-xfer.rst

Mint, transfer, and burn fungible tokens
========================================

Now you are ready to mint your new tokens. Before minting, look at the minting parameters in the JSON file below. You need to specify the owner, then the token data following with its tokenID, metadata location on IPFS with the URL that you got when you uploaded the metadata file, hash value of it, amount to be minted, and maximum supply number. In this case, it uses tokenID 01, and mint 110 initially with a maximum supply of 1000.

.. code-block:: JSON

    {
        "owner": {
            "Account": [
                "3AiAikC2v4H3Rv4L58oHvdX5N8LJoarsQwN3G7oNS7BoFsmt7N"
            ]
        },
        "tokens": [
            [
                "01",
                [
                    {
                        "url": "https://gateway.pinata.cloud/ipfs/QmZBrF6HuoN12HyAznyk7gwFpnefooDbfxq3JeKTWToL1W",
                        "hash": "c00a140ea66af09c343083c0e7ccc77d5e0670d90014569798801bffad637d15"
                    },
                    {
                        "amount": "110",
                        "max_supply": "1000"
                    }
                ]
            ]
        ]
    }

> Run the command below to call the contract ``mint()`` function. Remember that this is using the Concordium testnet node; if you are using your own node your command needs to be changed.

Reviewer: Please add how it needs to be changed (`--grpc-port` and `--grpc-ip` need to be changed (at least the latter, maybe not the former)

Reviewer 2: It shows in the deploy contract part for the rest, I don't think there is a need to point it out again.

Reviewer: Ok.

.. code-block:: console

    concordium-client contract update <YOUR-CONTRACT-INSTANCE> --entrypoint mint --parameter-json ../nft-artifacts/mint-params.json --schema dist/fungible/schema.bin --sender <YOUR-ACCOUNT> --energy 6000 --grpc-port 10000 --grpc-ip node.testnet.concordium.com

> Now you have minted your fungible tokens successfully.

Reviewer: ```suggestion
Now you have successfully minted your fungible tokens.
```

.. image:: images/mint-tokens.png
    :width: 100%
    :alt: dark console screen with result of running the command above

> Check the state to see the balance and other stuff. Run the command below to invoke the ``view()`` function.

Reviewer: "other stuff"?

Reviewer 2: Oops, sorry the blog language. I might be gotten distracted. "Let's check the state to see the balance, circulating supply, and maximum supply."

.. code-block:: console

>     concordium-client contract invoke <YOUR-CONTRACT-INSTANCE> --entrypoint view --schema dist/fungible/schema.bin --grpc-port 10000 --grpc-ip node.testnet.concordium.com

Reviewer: If you embedded the schema in the module these `--schema` options would not be needed. And the command would look cleaner.

This is also what users should do, so I recommend changing that.

Mint 500 more by adjusting the JSON file accordingly.

.. image:: images/mint-more.png
    :width: 100%
    :alt: dark console screen with result of running the command above

This was successful. Check the state once more by calling the ``view()`` function.

.. image:: images/invoke-after-mint-more.png
    :width: 100%
    :alt: dark console screen with result of running the command above

Transfer tokens
===============

Now you will test transferring some tokens to another account and checking the balances and the max/circulating supplies. Create a JSON file like the one below to set the transfer parameters. Transfer **11** tokens with the ID **01** to an account.

.. code-block:: JSON

    [
        {
            "token_id": "01",
            "amount": "11",
            "from": {
                "Account": [
                    "3AiAikC2v4H3Rv4L58oHvdX5N8LJoarsQwN3G7oNS7BoFsmt7N"
                ]
            },
            "to": {
                "Account": [
                    "4W59joBQavXqbyiKbPgBS7teZHDtEXsqx2BFUvPo2AzY3yPzeH"
                ]
            },
            "data": ""
        }
    ]

> Run the command below to invoke the transfer. Remember that this is using the Concordium testnet node; if you are using your own node your command needs to be changed.

Reviewer: As above.

.. code-block:: console

    concordium-client contract update <YOUR-INDEX> --entrypoint transfer --parameter-json ../token-artifacts/transfer.json --schema dist/fungible/schema.bin --sender <YOUR-ACCOUNT> --energy 6000 --grpc-port 10000 --grpc-ip node.testnet.concordium.com

Check the state one more time to see that the transfer worked as expected.

.. image:: images/invoke-after-xfer.png
    :width: 100%
    :alt: dark console screen with result of running the command above

Burn tokens
===========

> Finally, burn some tokens. Basically, the amount will be deducted from the owner's account when they call.

Reviewer: ```suggestion
Finally, burn some tokens. The amount will be deducted from the owner's account when the tokens are burned.
```
